src/comm/ModKqueue.cc

```diff
@@ -168,7 +168,7 @@ Comm::SetSelect(int fd, unsigned int type, PF * handler, void *client_data, time
 {
     fde *F = &fd_table[fd];
     assert(fd >= 0);
-    assert(F->flags.open);
+    assert(F->flags.open || (!handler && !client_data));
```

Comment:
Here and in similar cases, should we require a zero timeout as well?

Reply:
I don't think so. SetSelect supports changing the timeout when only FD number itself is known/knowable.

Reply:
It is difficult for me to relate your explanation to my question. Are you implying that it is OK for a SetSelect() caller to request a non-zero timeout on a closed file descriptor? And it is OK to schedule timeouts on a closed descriptor because the caller only has FD number and does not want to check whether the descriptor is still open??

Reply:
Done at 168f805.

Reply:
I mean we should not make Squid halt if that mistake is made. It is not actively harmful like setting a handler.

Reply:
> I mean we should not make Squid halt if that mistake is made.

Before this PR, Squid _was_ killed if somebody made the mistake of setting a timeout on a closed descriptor. This PR is not changing those deaths. This PR _relaxes_ the assert. My suggestion (implemented by Eduard) was to limit that relaxation to a very specific use case. ResetSelect() is the only (known) user matching that use case.